### DISCUSSION TITLE
Migrate to a new Inline Completion API

### DISCUSSION THREAD

**Discussion by BarracudaPff** [community/community#116679] [Copilot Conversations]
### Select Topic Area

General

### Body

Hi!
For a long time, IntelliJ Platform lacked a built-in API to display inline completion (aka gray text completion). As a result, many plugins, including Copilot, implemented it from scratch. In the latest 2023.3 release, the IntelliJ platform added an API for inline completion.

This new API allows developers to design plugins that enable inline completion, simplifying the process considerably. The primary benefit of this new API is that it enables multiple plugins to operate simultaneously, effectively enhancing overall system integration and compatibility.

The new built-in API is not compatible with implementations of other plugins, potentially causing two independent gray texts within the editor. We kindly invite you to explore the possibility of integrating our new API into your plugin at your earliest convenience. Your collaboration is crucial in achieving better synchronization between plugins for inline completion with the IntelliJ platform, thereby boosting the overall user experience.

Should you encounter any limitations or require additional features, please do not hesitate to let us know. We’re committed to ensuring this new API meets your requirements and improves the overall IntelliJ environment.

Here is [our issue tracker](https://youtrack.jetbrains.com/issues/IDEA?q=%23%7BEditor.%20Code%20Completion.%20Inline%7D).

Please migrate to the new API, you can find the entry points below:

- The API is located in the package [com.intellij.codeInsight.inline.completion](https://github.com/JetBrains/intellij-community/tree/master/platform/platform-impl/src/com/intellij/codeInsight/inline/completion).
    
- The [InlineCompletionProvider](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/codeInsight/inline/completion/InlineCompletionProvider.kt) is the main class responsible for providing inline completion suggestions. It responds to various [events](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/codeInsight/inline/completion/InlineCompletionEvent.kt), and for each one, it indicates whether it is ready to provide any suggestions for that event. There are default events: typing, changes in lookup, and direct call. If you require more events, you can create your own. There is a [version of the provider](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/codeInsight/inline/completion/DebouncedInlineCompletionProvider.kt) that uses delayed execution to avoid wasting resources on non-meaningful user inputs.
    
- For lower-level interaction, there's the [InlineCompletionHandler](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/codeInsight/inline/completion/InlineCompletionHandler.kt). You can attach your listeners to it in order to monitor what happens with the suggestions generated. Furthermore, you can send your custom events via InlineCompletionHandler.
    
- The provider yields a flow of [InlineCompletionElements](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/codeInsight/inline/completion/elements/InlineCompletionElement.kt) that need to be rendered. Because the flow is lazy, this API supports data streaming, allowing for the gradual appearance of elements. Each element features an [interface](https://github.com/JetBrains/intellij-community/blob/e4bc8f42efb7d244c8098f0d259352d90864a72a/platform/platform-impl/src/com/intellij/codeInsight/inline/completion/elements/InlineCompletionElement.kt#L25) that demonstrates how to render it. [The standard rendering](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/codeInsight/inline/completion/elements/InlineCompletionGrayTextElement.kt) of gray text is already provided. Should you require more complex elements, you can easily achieve this by extending the interface as described above.
    
- The API also supports [reaction to overtyping](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/codeInsight/inline/completion/suggestion/InlineCompletionSuggestionUpdateManager.kt), meaning instances where the user types characters that match the suggestion. Additionally, you can add a custom [insertion handler](https://github.com/JetBrains/intellij-community/blob/master/platform/platform-impl/src/com/intellij/codeInsight/inline/completion/InlineCompletionInsertHandler.kt) if you need to perform specific actions following element insertion.
    
- In 2023.3, support for only one suggestion per event is available out of the box. Support for multiple suggestions enabled in the 2024.1 release. However, you can also incorporate this functionality through custom events as detailed earlier.
    
As we discussed earlier, it should cover all your features with current implementation and even more.

Thank you in advance!

**Comment by thomasmuders-imtf** [comment]
Yes please update this as it breaks the new feature "Full line completion" in JetBrains IDE release 2024.1. This should be done ASAP.

**Comment by LukaDivjakovic** [comment]
Hello, I am supposed to use this API for a task in the process of getting an internship at JetBrains, but all the links are broken. Could you point me toward finding how to use this API?
